crates/dbs-virtio-devices/src/device.rs

```diff
@@ -0,0 +1,582 @@
+/// has device memory, such as DAX of virtiofs, pmem.
+///
+/// insert_region function is used to solve the problem that the virtio device cannot
+/// find the host address corresponding to the guest address when reading the
+/// guest device memory.
+///
+/// For example, the guest application executes the following code:
+/// {
+///     // "dax_fd" is virtio-fs file that support dax
+///     // "no_dax_fd" is virtio-fs file that do not support dax
+///     void *dax_ptr = (void*)mmap(NUMM, 4096, PORT, MAP_SHARED, dax_fd, 0);
+///     write(no_dax_fd, dax_ptr, 4096);
+/// }
+/// dragonball will coredump.
+///
+/// This is because the virtiofs device cannot resolve the dax_ptr address
+/// when calling vm_as.get_slice(). There is no DAX region in vm_as. This
+/// trait inserts the virtio device memory region, such as DAX region, into
+/// vm_as. This trait should be implemented in VMM when creating virtio
+/// devices with device memory, because the virtio device does not have
+/// permission to change vm_as.
+pub trait VirtioRegionHandler: Send {
+    /// Insert GuestRegionMmap to vm_as & address_space.
+    fn insert_region(&mut self, region: Arc<GuestRegionMmap>) -> std::result::Result<(), Error>;
+}
+
+/// Trait for virtio devices to be driven by a virtio transport.
+///
+/// The lifecycle of a virtio device is to be moved to a virtio transport, which will then query the
+/// device. Once the guest driver has configured the device, `VirtioDevice::activate` will be called
+/// and all the events, memory, and queues for device operation will be moved into the device.
+/// Optionally, a virtio device can implement device reset in which it returns said resources and
+/// resets its internal.
+pub trait VirtioDevice<AS: GuestAddressSpace>: Send + AsAny {
+    /// The virtio device type.
+    fn device_type(&self) -> u32;
+
+    /// The maximum size of each queue that this device supports.
+    fn queue_max_sizes(&self) -> &[u16];
+
+    /// The maxinum size of control queue
+    fn ctrl_queue_max_sizes(&self) -> u16 {
+        0
+    }
+
+    /// The set of feature bits shifted by `page * 32`.
+    fn get_avail_features(&self, page: u32) -> u32 {
+        let _ = page;
+        0
+    }
+
+    /// Acknowledges that this set of features should be enabled.
+    fn set_acked_features(&mut self, page: u32, value: u32);
+
+    /// Reads this device configuration space at `offset`.
+    fn read_config(&mut self, offset: u64, data: &mut [u8]);
+
+    /// Writes to this device configuration space at `offset`.
+    fn write_config(&mut self, offset: u64, data: &[u8]);
+
+    /// Activates this device for real usage.
+    fn activate(&mut self, config: VirtioDeviceConfig<AS>) -> ActivateResult;
+
+    /// Optionally deactivates this device and returns ownership of the guest memory map, interrupt
+    /// event, and queue events.
+    // fn reset(&mut self) -> Option<VirtioDeviceConfig<AS>> {
```

Comment:
Why comment this function?

Reply:
This file is not used yet, this pr should not contain the file, it will be introduced in the next pr.